### DISCUSSION TITLE
Storage factory Help

### DISCUSSION THREAD

**Discussion by dannyakorede12** [Cyfrin/foundry-full-course-cu#2467] [Q&A] [answered]
Hello, Just concluding the storage factory section of this course and I've run into a bit of an issue with compiling my StorageFactory 

this is the code
` ` ` 
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19; 

import {SimpleStorage} from "./SimpleStorage.sol";

contract StorageFactory{

    SimpleStorage[] public listofSimpleStorageContracts;

    function createSimpleStorageContract () public {
       SimpleStorage newSimpleStorageContract = new SimpleStorage();
       listofSimpleStorageContracts.push(newSimpleStorageContract);
    }

    function sfStore(uint256 _simpleStorageIndex, uint256 _newSimpleStorageNumber) public {
        SimpleStorage mySimpleStorage = listofSimpleStorageContracts[_simpleStorageIndex];
        mySimpleStorage.store(_newSimpleStorageNumber);
    }

    function sfGet(uint256 _simpleStorageIndex) public view returns (uint256) {
    SimpleStorage mySimpleStorage = listofSimpleStorageContracts[_simpleStorageIndex];
    return mySimpleStorage.retrieve();
    }
    
}

` ` ` 

I've tried debugging it myself, but what I'm getting is that there isn't a retrieve function in my SimpleStorage contract although asides the retrieve function for the favoriteNumber i don't think it was anywhere else

I'm sure i've followed the course quite well so far and I seem to be at wits end.

I'd appreciate help with this. Thank You

**Comment by EngrPips** [comment]
Hello @dannyakorede12, You can see that both `sfStore` and `sfGet` takes an argument of `_simpleStorageIndex`, which is the index of the `simpleStorage` contract we want to interact with from the `SimpleStorage[] public listofSimpleStorageContracts`  which hold the list of `simpleStorage` contract that the `StorageFactory` contract has deployed. So the chain of the transaction is you call `sfStore` on the `StorageFactory` contract passing the `_simpleStorageIndex` argument to the function which specifies the index of the actual `simpleStorage` you want to interact with from the list of `simpleStorage` contract already deployed by the `StorageFactory` The `StorageFactory` gets out the `simpleStorage` contract and call the `Store` function on the `simpleStorage`  

If you go back to look at the `simpleStorage` contract, you will see that it has both `store` and `get` so we are only calling the `StorageFactory` so that it can call the  `simpleStorage` contract.

Please ask question if anything is confusing.

**Reply by dannyakorede12** [reply]
Hi, thanks for the response, but i'm kinda a bit confused but if i'm getting you right, what you said is to call the sfStore on the StorageFactory itself and not on the _simpleStorageIndex .

is that what you mean?

**Reply by EngrPips** [reply]
What I am trying to say is that you should call the `sfGet` and `sfStore` functions on the `StorageFactory`, and then you will pass in the index of the `simpleStorage` contract you want it to call the `get` and `store` function on. I will advise that you take the lesson if things are very confusing.

**Reply by dannyakorede12** [reply]
Thanks once again for your response, and beleive me when i say, i've gone over this section severally and still don't seem to know where i missed it.

my understanding from all this is there might be a problem with the simple storage contract itself, so here is the code for that. please let me know if there's anything i missed that's affecting the storage factory code.

`contract SimpleStorage {
    
    uint256 myFavoriteNumber;

    
    struct Person{
        uint256 favoriteNumber;
        string name;
    }


    Person[] public listofPeople; 
    

    mapping(string => uint256) public nameTofavoriteNumber;
    
    function store(uint256 _favoriteNumber) public{
        myFavoriteNumber = _favoriteNumber;
    }

    function retreive() public  view returns (uint256){
        return myFavoriteNumber;
    }
    

    function addPerson(string memory _name, uint256 _favoriteNumber) public {
        listofPeople.push(Person(_favoriteNumber, _name));
           nameTofavoriteNumber[_name] = _favoriteNumber;     
    }
 `

**Reply by EngrPips** [reply, marked as answer]
The problem is you misspelled `retrieve` in this `SimpleStorage`  contract. you spelled it as `retreive` instead of `retrieve`. Please correct the spelling and retry.

**Reply by dannyakorede12** [reply]
Wow, so crazy how a spelling error has been giving me issues for months!!! i think going forward i would have to pay more attention to that. 

Thanks for the fresh pair of eyes, got too used to looking at it that i missed it. it worked!!

**Reply by EngrPips** [reply]
I could only spot that because I have also made the same mistake before. Experience is the best teacher, no doubt. Now you will always keep spelling mistakes in mind whenever you are coding.

**Reply by dannyakorede12** [reply]
that's great. Thanks a lot.
I sure will keep them in mind

**Reply by EngrPips** [reply]
Indeed, Keep crushing it.
